staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,700 @@
+	qsCompleter, err := qscOfPL(meal.cfgCtl.queueSetFactory, nil, proto.Name, &proto.Spec, requestWaitLimit)
+	if err != nil {
+		// This can not happen because proto is one of the mandatory
+		// objects and these are not erroneous
+		panic(err)
+	}
+	meal.newPLStates[proto.Name] = &priorityLevelState{
+		config:      proto.Spec,
+		qsCompleter: qsCompleter,
+	}
+	if proto.Spec.Limited != nil {
+		meal.shareSum += float64(proto.Spec.Limited.AssuredConcurrencyShares)
+	}
+	return
+}
+
+func (cfgCtl *configController) Match(rd RequestDigest) (string, *fctypesv1a1.FlowDistinguisherMethod, string, StartFunction) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	var fs *fctypesv1a1.FlowSchema
+	for _, fs = range cfgCtl.flowSchemas {
+		if matchesFlowSchema(rd, fs) {
+			plName := fs.Spec.PriorityLevelConfiguration.Name
+			plState := cfgCtl.priorityLevelStates[plName]
+			if plState.config.Type == fctypesv1a1.PriorityLevelEnablementExempt {
+				klog.V(7).Infof("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, immediate", rd, fs.Name, fs.Spec.DistinguisherMethod, plName)
+				return fs.Name, fs.Spec.DistinguisherMethod, plName, nil
+			}
+			plState.numPending++
+			matchID := &plName
+			klog.V(7).Infof("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID)
+			startCalls := 0
+			startFn := func(ctx context.Context, hashValue uint64) (execute bool, afterExecution func()) {
+				startCalls++
+				klog.V(7).Infof("For matchID=%p, startFn(ctx, %v) startCalls:=%d", matchID, hashValue, startCalls)
+				if startCalls > 1 {
+					panic(fmt.Sprintf("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p startCalls:=%d", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID, startCalls))
+				}
+				req := func() fq.Request {
+					cfgCtl.lock.Lock()
+					defer cfgCtl.lock.Unlock()
+					plState.numPending--
+					req, idle1 := plState.queues.StartRequest(ctx, hashValue, rd.RequestInfo, rd.User)
+					if idle1 {
+						cfgCtl.maybeReapLocked(plName, plState)
+					}
+					return req
+				}()
+				if req == nil {
+					klog.V(7).Infof("For matchID=%p, startFn(ctx, %v): reject", matchID, hashValue)
+					return false, func() {}
+				}
+				exec, idle2, afterExec := req.Wait()
+				execID := &exec
+				if idle2 {
+					cfgCtl.maybeReap(plName)
+				}
+				klog.V(7).Infof("For matchID=%p, startFn(..) => exec=%v, execID=%p", matchID, exec, execID)
+				afterCalls := 0
+				return exec, func() {
+					afterCalls++
+					klog.V(7).Infof("For matchID=%p, execID=%p, after() calls:=%d", matchID, execID, afterCalls)
+					if afterCalls > 1 {
+						panic(fmt.Sprintf("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p, execID=%p, afterCalls:=%d", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID, execID, afterCalls))
+					}
+					idle3 := afterExec()
+					if idle3 {
+						cfgCtl.maybeReap(plName)
```

Comment:
maybeReapLocked? `cfgCtl.lock` will be deadlock

Reply:
this anonymous function is also called by the client without `cfgCtl` locked.